cowprotocol/cowswap
# Navigate with widget awareness

# Summary

Follow up on this PR https://github.com/cowprotocol/cowswap/pull/4470
Fixes #4383   

This second PR makes it more robust, so we not only stop polluting the history when the widget changes the parameters but in any part of the app.

In order to do this, I do 2 things

## Custom useNavigate hook
I implemented a custom version of the react-router useNavigate hook which will be widget aware

## ES-lint rule to enforce that
It seems very easy mistake to use the react router hook instead of ours, so I added a ES-lint rule that will yell at you if you do.

<img width="1492" alt="Screenshot at May 22 19-07-02" src="https://github.com/cowprotocol/cowswap/assets/2352112/796f6987-924a-4bf5-b400-1e72b144cbf5">

## Bonus
I added for convenience a script to the package.json to run `nx` things

I just find super weird to type so many words to launch the lint: `npx nx run ....` so now you just do `yarn nx` 

For example, to test this PR you will use `yarn nx cowswap-frontend:lint`

<img width="639" alt="Screenshot at May 22 19-20-36" src="https://github.com/cowprotocol/cowswap/assets/2352112/c1f9d06c-08d0-4e66-ad8e-878794b4d785">

## Not included
This is done for the all programmatic navigations, however we might want to adapt the <Link /> too.

Because we don't have links in the widget I didn't do it for now, as it might be overkill, but we could do a similar approach here:
- Make our own component
- Add the lint rule

```ts
import { Link as OriginalLink } from 'react-router-dom';

export function Link({ to, replace = isWidget, ...props }) {
  return <OriginalLink to={to} replace={replace} {...props} />;
}
```

## Test
- Open configurator https://widget-configurator-git-navigate-with-widget-awareness-cowswap.vercel.app/
- Change the colors of the PAPER a few times
- Navigate back, it should go to wherever you were before CoW Swap (or the home of your browser)

These steps would not work in here, where going back would be changing the colors back:
https://widget.cow.fi
